### PR TITLE
ZipBuilder: Get dependencies from the Podfile.lock

### PR DESCRIPTION
**Pull request by paulb777**
This is a speedup and prerequisite for supporting local podspec's.

See embedded comments below for more detail on the implementation.

#no-changelog

### CODE REVIEW

File: ZipBuilder/Sources/ZipBuilder/ZipBuilder.swift [hunk 1]
```diff
@@ -437,7 +443,7 @@ struct ZipBuilder {
       result += "- \(framework).framework\n"
     }
 
-    result += "\n"
+    result += "\n" // Necessary for Resource message to print properly in markdown.
```
Comment (paulb777) [inline ZipBuilder/Sources/ZipBuilder/ZipBuilder.swift]: These changes will go away when #4450 lands

File: ZipBuilder/Sources/ZipBuilder/ZipBuilder.swift [hunk 2]
```diff
@@ -345,7 +351,7 @@ struct ZipBuilder {
   /// - Throws: Various FileManager errors in case the copying fails, or an error if the framework
   //            doesn't exist in `frameworkLocations`.
   @discardableResult
-  func copyFrameworks(fromPods installedPods: [CocoaPodUtils.PodInfo],
+  func copyFrameworks(fromPods installedPods: [String],
```
Comment (paulb777) [inline ZipBuilder/Sources/ZipBuilder/ZipBuilder.swift]: Only need the names

File: ZipBuilder/Sources/ZipBuilder/ZipBuilder.swift [hunk 3]
```diff
@@ -125,10 +125,12 @@ struct ZipBuilder {
   /// - Parameter podsToInstall: All pods to install.
   /// - Returns: Arrays of pod install info and the frameworks installed.
   func buildAndAssembleZip(podsToInstall: [CocoaPodUtils.VersionedPod]) ->
-    ([CocoaPodUtils.PodInfo], [String: [URL]]) {
+    ([String: CocoaPodUtils.PodInfo], [String: [URL]]) {
     // Remove CocoaPods cache so the build gets updates after a version is rebuilt during the
     // release process.
-    CocoaPodUtils.cleanPodCache()
+    if LaunchArgs.shared.updatePodRepo {
```
Comment (paulb777) [inline ZipBuilder/Sources/ZipBuilder/ZipBuilder.swift]: Diff will go away when #4449 lands

File: ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift [hunk 4]
```diff
@@ -25,19 +25,20 @@ public enum CocoaPodUtils {
...
 
   /// Information associated with an installed pod.
   public struct PodInfo {
-    let versionedPod: VersionedPod
+    /// The version of the generated pod.
+    let version: String
 
-    /// The location of the pod on disk.
-    var installedLocation: URL
+    /// The pod dependencies
```
Comment (ryanwilson) [inline ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift]: Nit: period.

File: ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift [hunk 5]
```diff
@@ -54,23 +55,8 @@ public enum CocoaPodUtils {
...
-      fatalError("Could not list the pod cache in \(dir), the command exited with \(code). Try " +
-        "running in Terminal to see what's wrong.")
-    case let .success(output):
-      return parsePodsCache(output: output.components(separatedBy: "\n"))
-    }
-  }
-
   /// Gets metadata from installed Pods. Reads the `Podfile.lock` file and parses it.
-  public static func installedPodsInfo(inProjectDir projectDir: URL) -> [PodInfo] {
+  public static func installedPodsInfo(inProjectDir projectDir: URL) -> [String: PodInfo] {
```
Comment (ryanwilson) [inline ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift]: Can you mention what the key is in the documentation?

File: ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift [hunk 6]
```diff
@@ -80,33 +66,16 @@ public enum CocoaPodUtils {
...
+    return loadPodInfoFromPodfileLock(contents: podfileLock)
   }
 
   /// Install an array of pods in a specific directory, returning an array of PodInfo for each pod
   /// that was installed.
   @discardableResult
   public static func installPods(_ pods: [VersionedPod],
                                  inDir directory: URL,
-                                 customSpecRepos: [URL]? = nil) -> [PodInfo] {
+                                 customSpecRepos: [URL]? = nil) -> [String: PodInfo] {
```
Comment (ryanwilson) [inline ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift]: Nit: documentation is now out of date

File: ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift [hunk 7]
```diff
@@ -148,31 +117,65 @@ public enum CocoaPodUtils {
...
 
-  /// Load versions of installed Pods from the contents of a `Podfile.lock` file.
+  /// Load installed Pods from the contents of a `Podfile.lock` file.
   ///
   /// - Parameter contents: The contents of a `Podfile.lock` file.
-  /// - Returns: A dictionary with names of the pod for keys and a string representation of the
-  ///            version for values.
-  public static func loadVersionsFromPodfileLock(contents: String) -> [String: String] {
-    // This pattern matches a framework name with its version (two to three components)
+  /// - Returns: An array of PodInfo structs.
```
Comment (ryanwilson) [inline ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift]: Nit: this is incorrect, it's a dictionary

File: ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift [hunk 8]
```diff
@@ -148,31 +117,65 @@ public enum CocoaPodUtils {
...
+    var currentPod: String?
+    for line in contents.components(separatedBy: .newlines) {
+      if line.starts(with: "DEPENDENCIES:") {
+        break
+      }
+      if let (pod, version) = detectVersion(fromLine: line, matching: podRegex) {
+        let corePod = pod.components(separatedBy: "/")[0]
+        currentPod = corePod.trimmingCharacters(in: quotes)
+        pods[currentPod!] = version
+      } else if let curPod = currentPod {
```
Comment (ryanwilson) [inline ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift]: Nit: I believe we should be able to do `else if let currentPod = currentPod` which gives you an unwrapped version of the same name, it makes it a bit more readable I find.

File: ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift [hunk 9]
```diff
@@ -212,6 +215,33 @@ public enum CocoaPodUtils {
     }
   }
 
+  /// Get all transitive pod dependencies for a pod.
+  /// - Returns: An array of Strings
+  static func transitivePodDependencies(for podName: String,
+                                        in installedPods: [String: PodInfo]) -> [String] {
+    var newDeps = Set([podName])
+    var returnDeps = Set<String>()
+    repeat {
```
Comment (ryanwilson) [inline ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift]: I forgot about `repeat-while`, nice!

File: ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift [hunk 10]
```diff
@@ -212,6 +215,33 @@ public enum CocoaPodUtils {
     }
   }
 
+  /// Get all transitive pod dependencies for a pod.
+  /// - Returns: An array of Strings
```
Comment (ryanwilson) [inline ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift]: What is the content of the strings, pod names?

File: ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift [hunk 11]
```diff
@@ -212,6 +215,33 @@ public enum CocoaPodUtils {
...
+        foundDeps.formUnion(Set(childDeps))
+      }
+      newDeps = foundDeps.subtracting(returnDeps)
+      returnDeps.formUnion(newDeps)
+    } while newDeps.count > 0
+    return Array(returnDeps)
+  }
+
+  /// Get all transitive pod dependencies for a pod.
+  /// - Returns: An array of VersionedPod
```
Comment (ryanwilson) [inline ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift]: Nit: An array of dependencies with versions for a given pod.

File: ZipBuilder/Sources/ZipBuilder/ZipBuilder.swift (hunk 2 above)
Comment (ryanwilson) [reply ZipBuilder/Sources/ZipBuilder/ZipBuilder.swift]: Can you please change the comment to `Names of all the Pods installed...` or something like that?

File: ZipBuilder/Sources/ZipBuilder/ZipBuilder.swift [hunk 12]
```diff
@@ -537,41 +545,31 @@ struct ZipBuilder {
   /// directory.
   ///
   /// - Parameter projectDir: The directory containing the Podfile.lock file of installed pods.
-  private func validateExpectedVersions(inProjectDir projectDir: URL) {
+  private func validateExpectedVersions(installedPods: [String: CocoaPodUtils.PodInfo]) {
```
Comment (ryanwilson) [inline ZipBuilder/Sources/ZipBuilder/ZipBuilder.swift]: Nit: documentation is now out of date

File: ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift (hunk 4 above)
Comment (paulb777) [reply ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift]: Done

File: ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift (hunk 5 above)
Comment (paulb777) [reply ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift]: Added a func doc block.

File: ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift (hunk 6 above)
Comment (paulb777) [reply ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift]: Fixed.

File: ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift (hunk 7 above)
Comment (paulb777) [reply ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift]: Fixed.

File: ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift (hunk 8 above)
Comment (paulb777) [reply ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift]: Thanks!

File: ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift (hunk 10 above)
Comment (paulb777) [reply ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift]: Yep. Clarified.

File: ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift (hunk 11 above)
Comment (paulb777) [reply ZipBuilder/Sources/ZipBuilder/CocoaPodUtils.swift]: Done.

File: ZipBuilder/Sources/ZipBuilder/ZipBuilder.swift (hunk 2 above)
Comment (paulb777) [reply ZipBuilder/Sources/ZipBuilder/ZipBuilder.swift]: Done.

File: ZipBuilder/Sources/ZipBuilder/ZipBuilder.swift (hunk 12 above)
Comment (paulb777) [reply ZipBuilder/Sources/ZipBuilder/ZipBuilder.swift]: Fixed.

**Pull request merged**